lisa/sut_orchestrator/hyperv/source.py

```diff
@@ -0,0 +1,190 @@
+            str(dest_path) if destination_is_dir else str(dest_path.parent)
+        )
+
+        self._log.debug(f"Copying {local_path} to server")
+        server.shell.copy(local_path, dest_path)
+        self._log.debug("Finished copying.")
+
+        return [dest_path]
+
+    def _download_zipped_file(
+        self, local_path: Path, destination: Optional[str], server: RemoteNode
+    ) -> List[PureWindowsPath]:
+        destination_is_dir = destination and (
+            destination.endswith("/") or destination.endswith("\\")
+        )
+
+        defined_destination = None
+        if destination:
+            if PureWindowsPath(destination).is_absolute():
+                defined_destination = PureWindowsPath(destination)
+            else:
+                defined_destination = PureWindowsPath(server.working_path / destination)
+
+        tmp_dest = PureWindowsPath(
+            server.working_path / "zipped_sources_tmp" / local_path.name
+        )
+
+        server.tools[Mkdir].create_directory(str(tmp_dest.parent))
+
+        server.shell.copy(local_path, tmp_dest)
+
+        # if the destination is a directory, directly unzip the file(s) into it
+        extraction_path = self._unzip_file(
+            server, tmp_dest, defined_destination if destination_is_dir else None
+        )
+
+        extracted_files = server.tools[Ls].list(str(extraction_path))
+
+        if destination_is_dir:
+            # the files are already in the destination directory
+            return [extraction_path.joinpath(f) for f in extracted_files]
+
+        if not defined_destination:
+            final_destination = PureWindowsPath(server.working_path / "sources")
+            server.tools[Mkdir].create_directory(str(final_destination))
+            server.tools[PowerShell].run_cmdlet(
+                f"Copy-Item -Path '{extraction_path}\\*'"
+                f"-Destination '{final_destination}\\' -Recurse"
+            )
+            server.tools[Rm].remove_file(f"{extraction_path}\\*")
+            return [(final_destination / f) for f in extracted_files]
+
+        # if the destination is not a directory and it is defined in the runbook,
+        # then it has to be a single file.
+        assert len(extracted_files) == 1
+        final_destination = PureWindowsPath(
+            server.working_path / "sources" / extracted_files[0]
+        )
+
+        self._log.debug(f"Copying {local_path} to server")
+        server.tools[Cp].copy(extraction_path / extracted_files[0], final_destination)
+        self._log.debug("Finished copying.")
+
+        server.tools[Rm].remove_file(extracted_files[0])
+
+        return [final_destination]
+
+    def _unzip_file(
+        self,
+        server: RemoteNode,
+        zipped_file: PureWindowsPath,
+        destination_dir: Optional[PureWindowsPath] = None,
+    ) -> PureWindowsPath:
+        extraction_path = zipped_file.parent.joinpath(f"extracted-{uuid.uuid4()}")
```

Comment:
use the default path in the zip file. So the unzipped path is predictable.

Reply:
This is a temporary staging path for unzip. After unzip, it is copied to the final destination.

Reply:
If as per the above comment, we assume destination is always a directory, this may be simplified. I'll see.